cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,388 @@
+package v1alpha1
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+)
+
+var RolloutClock = clock.Clock(clock.RealClock{})
+var maxTimeDuration = time.Duration(math.MaxInt64)
+
+type RolloutStatus int
+
+const (
+	// resource desired status is not applied yet
+	ToApply RolloutStatus = iota
+	// resource desired status is applied and last applied status is not updated
+	Progressing
+	// resource desired status is applied and last applied status is succeed
+	Succeed
+	// resource desired status is applied and last applied status is failed
+	Failed
+	// when the rollout status is progressing or failed and the status remains for longer than the timeout
+	// then the status will be set to timeout
+	TimeOut
+	// skip rollout on this cluster
+	Skip
+)
+
+// Return the rollout status on each managed cluster
+type ClusterRolloutStatusFunc func(clusterName string) ClusterRolloutStatus
+
+type ClusterRolloutStatus struct {
+	// cluster name
+	// required field
+	Name string
+	// cluster group key
+	// optional field
+	GroupKey clusterv1beta1.GroupKey
+	// rollout status
+	// required field
+	Status RolloutStatus
+	// the last transition time of rollout status
+	// this is used to calculate timeout for progressing and failed status
+	// optional field
+	LastTransitionTime *time.Time
+}
+
+type RolloutResult struct {
+	// clusters to rollout
+	ClustersToRollout map[string]ClusterRolloutStatus
+	// clusters that are timeout
+	ClustersTimeOut map[string]ClusterRolloutStatus
+}
+
+// +k8s:deepcopy-gen=false
+type RolloutHandler struct {
+	// placement decision tracker
+	pdTracker *clusterv1beta1.PlacementDecisionClustersTracker
+}
+
+func NewRolloutHandler(pdTracker *clusterv1beta1.PlacementDecisionClustersTracker) *RolloutHandler {
+	if pdTracker == nil {
+		return nil
+	}
+
+	return &RolloutHandler{
+		pdTracker: pdTracker,
+	}
+}
+
+// The input is a duck type RolloutStrategy and a ClusterRolloutStatusFunc to return the rollout status on each managed cluster.
+// Return the strategy actual take effect and a list of cluster that needs to update with current state.
+func (r *RolloutHandler) GetRolloutCluster(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, RolloutResult, error) {
```

Comment:
What's the expected behavior for the caller to know when a cluster or cluster group has timed out?

It seems to me that somehow returning a context with a deadline for every potential timeout would allow the caller to poll the channel of the context to know if a timeout occurred and then verify if the rollout finished before that timeout.

Reply:
@serngawy mentions that some policies might keep in progressing status on some clusters for a long time, don't want them to block upgrading others, and also need to know which are timeout, right?

Reply:
What I'm getting at is how is the controller to know when a timeout has occurred? I'd like this to be event driven as opposed to continuously polling `GetRolloutCluster`. My idea was a context with a timeout could be returned so the controller can be notified when the timeout has occurred.

Reply:
Try to return the timeout time in the ClusterRolloutStatus , so that the consumer controller can use this time to requeue the related resource like `syncCtx.Queue().AddAfter(key, duration)`
 
```
type ClusterRolloutStatus struct {
	// cluster group key
	// optional field
	GroupKey clusterv1beta1.GroupKey
	// rollout status
	// required field
	Status RolloutStatus
	// the last transition time of rollout status
	// this is used to calculate timeout for progressing and failed status
	// optional field
	LastTransitionTime *metav1.Time
	// the timeout time when status is progressing or failed
	// optional field
	TimeOutTime *metav1.Time
}
```